### ISSUE TITLE
[ILM] Data retention UX

### ISSUE THREAD

**Issue by jloleysens** [elastic/elasticsearch#66097]
## Summary

The current ILM API does not fit the new data retention UI/UX which creates issues for users in **reasoning about time** data has spent in a phase (and by extension a data tier). This impacts understanding of cluster resource use and cost.

The crux of the issue is **minimum age** vs **duration**.

## The problem

The current ILM API presents timing configuration in terms of minimum age. It is made to express the following:

* Data should be a minimum of 5 days old before moving to the warm phase
* Data should be a minimum of 15 days old before it is moved to the cold phase
* Should be a minimum of 85 days old before it is deleted

The “5”, “15” and “85” days are all from the time of index creation (ignoring complexities of rollover for now). The difficulty for users of the API is that configuring timing in this way requires some math to answer the question: **How long is my data in the warm phase?** This question has become more important since the formalisation of data tiers because phase determines where data is stored and how long.

## Solutions

Not exhaustive, but here are the most obvious solutions to this problem:

### 1. Change the ES API

The ILM API should support specifying timing in terms of **durations**, expressing the same policy as above:

* 5 days in hot phase
* 10 days in warm phase
* 70 days in cold phase, then delete

Per the problem statement.

This will require work to make ILM understand durations, but the _massive benefit_ is the UI and ES (API and UI users) will speak the same language. There is no need for any conversion or calculations to represent the data retention UX we want.

### 2. Don’t change the API, handle only in the front end

In order to express durations in the UI a conversion layer needs to be implemented doing:

* “5 days in Hot” -> “5 days old before moving to warm”
* “10 days in Warm” -> “15 days old before moving to cold”
* “70 days in Cold then delete” -> “85 days old before deleted”

The problem is this diverges the UI/UX from the API in a way that not only creates a maintenance burden on the UI/UX (code to write, test and maintain) but a different way of reasoning about timing will generate a policy users, and those responsible for supporting an ES cluster, will not understand - the crux of the problem being language mismatch that leads to miscommunication.

Additionally, the ability to specify timing in different units (seconds, hours, days) creates more conversion complexity and will force the UI to use the smallest selected time unit so that all time units are the same and we can do math for conversion. For pre-existing policies (pre new UI) the UI needs to decode these numbers back to “probably” the time unit the user input, introducing some potential for unexpected results in converting an already created policy into this converted model. This holds also for policies created using the new UI, but less so.

All in all, this is _quite problematic_.

## Proposal and request

Since solution 2 is fraught with issues, it looks like solution 1 is what we want, but perhaps something in between is most feasible.

Assuming the time investment estimation for solution 1 is outside of the timeframe we accept for delivering the new UX, the UI can for a time adopt solution 2 with the understanding that solution 1 is the one we will eventually have.

A synthesis might require some other API changes like adding a metadata field to ILM policies so that the UI can track which policies use the new UX, and which use the old so that we can reduce the amount of changes needed for existing policies.

All of this requires estimating the amount of work and time required to make a change like this. Assistance with thinking through this design and potentials for addressing the problem would be greatly appreciated.

## Additional notes

* How rollover impacts the thinking on durations and reasoning about how long and where data is needs to be explored

CC @jasontedor @jethr0null @dakrone @andreidan

**Comment by dakrone** [comment]
I think this misses the potential solution of changing the UI to deal with age rather than duration (basically solution 1, but flipped).

I still prefer that as an option as I think it's easier to think of age because it references an absolute point (either rollover or index creation time) rather than requiring a user to combine durations (we force the user to calculate their retention by adding durations: "let's see, 15 days in hot plus 30 in warm plus 40 in cold means deleting at ~85 days").

**Comment by andreidan** [comment]
I don't think we can ignore the complexities of rollover. The phase transition timings are centred around rollover and a hot phase without rollover is at most of a niche case (if that). 

I believe this is a presentation layer issue. I'd say the API is "presentationless" and interpreting the results and formatting them to different formats is something the UI would have to do, so I would prefer solution 2.

With regards to how the users will have to reason about the policies, I'd argue that the current age based timings are closer to what the users have to think about when configuring the policies ie. "data needs to be stored for 90 days" translates into a `min_age` of `90` in the delete phase, as opposed to having to do math whilst configuring each phase to make sure the sum of all other `min_ages` is get to 90. IMO the current age based timings amount to a friendlier UX. I believe this exact use case is visible in the solution 1 example in the description, namely the index would be deleted after `85` days which is not immediately visible.

The drawbacks of solution 2 imply a user will keep going back and forth between the API and the UI (ie. configure in the UI, go and check the API, then come back in the UI? ). Is this a valid user story? 
If it is, would we be able to provide both terminologies? (ie. be able to switch from retention /phase lingo to index age based retention display?) Or additional information when configuring the policy in the UI? (a "time translation" indicator?)

With regards to someone monitoring the policies using the explain API/ilm-history index, we show the index age so I believe there wouldn't necessarily be a mismatch between the policy definition and the debugging/monitoring information.

**Comment by jloleysens** [comment]
Thanks both for reading the issue and responding!

@dakrone 

> I still prefer that as an option as I think it's easier to think of age because it references an absolute point.

I think we have driven this to a conclusion in offline discussion: **absolute timing provides a more desirable UX** and so I think this issue has been adequately addressed, closing this for now. We can support editing absolute time with just presenting relative time as an indication for how long data will be in a phase.

I thought I would try to clarify some of the points I made originally, but feel free to ignore since they are no longer relevant given the above decision.

@andreidan

> I don't think we can ignore the complexities of rollover.

I think this is true. I was trying to focus the issue on just how we present timing for the purpose of thinking through the UX at different stack levels.

> The drawbacks of solution 2 imply a user will keep going back and forth between the API and the UI

Apologies that my description of the issue was not clear enough. There were two issues with solution 2 IMO, one was that UI users and API users, like support, would not use the same config for the same thing (one sees absolute the other sees relative) and the other was that the UI needed to mutate policy values just because the policy was changed in the UI which did not seem desirable to me - we need to present + edit relative time. Most of the challenge here is the edit part.
